File: src/praisonai-agents/praisonaiagents/agent/placement.py
"""One vocabulary for "where does this run?", shared by Agent, AgentFlow and AgentTeam.

Two questions get asked about placement and they have different answers:

``run_on=``
    Where the **whole agent** runs -- the model calls, the loop, the tools, all
    of it on someone else's machine. Only a *managed runtime* can answer this,
    because running the loop remotely means something has to host the loop.

``tools_run_on=``
    Where **the tools** run. Thinking stays on this machine; only the shell,
    the file writes and the code execution move. Any sandbox or compute place
    can answer this, because all it has to do is run a command.

Keeping the two words apart is the entire point. The old surface used one word
(``run_on=``) on ``AgentFlow``/``AgentTeam`` for the tools-only meaning, another
(``compute=``) on ``LocalAgent`` for exactly the same meaning, and a third
(``backend=``) for the whole-agent meaning -- so the same concept had two names
and one name had two concepts.

Because the two questions have disjoint answer sets, a wrong pairing is a typo
worth catching rather than a preference worth honouring. ``run_on="docker"``
cannot mean anything: Docker does not host agent loops. It gets a ``TypeError``
naming the parameter the user actually wanted, not a silent fallback.
"""

from __future__ import annotations

from typing import Any, Literal, NamedTuple, Optional, Union

# Runtimes that can host an entire agent loop. Resolved from the wrapper's
# entry-point registry when it is installed so third-party backends work with
# `run_on=` too; the literal is the floor, not the ceiling.
_FALLBACK_MANAGED = ("anthropic",)

# Static spellings so editors can autocomplete and type-checkers can reject a
# typo at the call site rather than at the first model turn. The runtime
# functions below stay the authority for validation -- these exist for tooling,
# which cannot call a function to learn a set.
# Keep in sync with _compute_bridge._PROVIDERS | _sandbox_adapter.SANDBOX_ONLY.
ManagedRuntime = Literal["anthropic"]

ToolPlace = Literal[
    "local", "subprocess", "sandlock", "docker", "e2b",
    "modal", "daytona", "flyio", "tenki", "novita", "ssh",
]


def managed_runtimes() -> tuple:
    """Names that ``run_on=`` accepts -- places that host a whole agent."""
    try:
        from praisonai.integrations.backend_registry import get_backend_registry

        names = tuple(get_backend_registry().list_all_names())
        if names:
            return tuple(sorted(set(names) | set(_FALLBACK_MANAGED)))
    except Exception:
        pass
    return _FALLBACK_MANAGED


def tool_places() -> list:
    """Names that ``tools_run_on=`` accepts -- places that can run a command.

    Includes the resolver's aliases. Validation runs before resolution, so
    leaving them out rejected spellings the resolver handles happily:
    ``tools_run_on="native"`` raised "not a known place" while
    ``resolve_compute("native")`` returned a working sandlock adapter.
    """
    try:
        from ..managed._compute_bridge import _ALIASES, available_providers

        return sorted(set(available_providers()) | set(_ALIASES))
    except Exception:
        return ["local", "docker"]


class Placement(NamedTuple):
    """The resolved answer to "where does this run?"."""

    whole: Optional[Any] = None   # managed runtime name/object, or None
    tools: Optional[Any] = None   # tools place name/object, or None


def _name_of(value: Any) -> Optional[str]:
    """The comparable name of a placement value, or None if it is an object.

    Objects (a built provider, a live ``SSHSandbox``) carry their own config and
    are never second-guessed -- only bare strings get routed by name.
    """
    return value.lower() if isinstance(value, str) else None


def _quote_list(names) -> str:
    return ", ".join(repr(n) for n in names)


def resolve_placement(
    owner: str,
    *,
    run_on: Any = None,
    tools_run_on: Any = None,
    backend: Any = None,
    supports_run_on: bool = True,
) -> Placement:
    """Validate placement arguments and return where things should run.

    ``owner`` is the class name, used only to make errors read naturally.
    ``supports_run_on`` is False for AgentFlow/AgentTeam, which can place tools
    but cannot (yet) hand a whole multi-agent run to a managed runtime.

    Raises:
        TypeError: on any combination that names two different places for the
            same thing, or a place that cannot do the job asked of it.
    """
    # ── a place that cannot do the job asked of it ───────────────────────────
    run_on_name = _name_of(run_on)
    tools_name = _name_of(tools_run_on)

    # The two lookups below feed only the branches guarded on run_on_name /
    # tools_name being set; the entry-point scan in tool_places() is the
    # dominant cost of Agent() construction, so skip both on the default
    # (no-placement) path rather than compute results that get discarded.
    if run_on_name is not None or tools_name is not None:
        managed = managed_runtimes()
        places = tool_places()
    else:
        managed = places = ()

    if run_on_name is not None and run_on_name not in managed:
        if run_on_name in places:
            raise TypeError(
                f"{owner}(run_on={run_on!r}) is not valid: run_on= places the "
                f"whole agent -- model calls, loop and tools -- on a managed "
                f"runtime, and {run_on_name!r} runs commands but cannot host an "
                f"agent loop.\n"
                f"  To run only the tools there:  {owner}(tools_run_on={run_on!r})\n"
                f"  Runtimes that host a whole agent: {_quote_list(managed)}"
            )
        raise TypeError(
            f"{owner}(run_on={run_on!r}) is not a known managed runtime.\n"
            f"  Runtimes that host a whole agent: {_quote_list(managed)}\n"
            f"  Places that can run tools:        {_quote_list(places)} "
            f"(pass those as tools_run_on=)"
        )

    if tools_name is not None and tools_name in managed and tools_name not in places:
        raise TypeError(
            f"{owner}(tools_run_on={tools_run_on!r}) is not valid: {tools_name!r} "
            f"hosts an entire agent, not individual tools.\n"
            f"  Did you mean:  {owner}(run_on={tools_run_on!r})"
        )

    # A typo here used to survive construction and only surface on the first
    # model turn -- after the prompt was built and, for a real run, after the
    # API spend. run_on= validates at the call site and YAML validates at load
    # time; this is the same standard for the third spelling.
    if tools_name is not None and tools_name not in places:
        raise TypeError(
            f"{owner}(tools_run_on={tools_run_on!r}) is not a known place.\n"
            f"  Places that can run tools: {_quote_list(places)}\n"
            f"  Runtimes that host a whole agent: {_quote_list(managed)} "
            f"(pass those as run_on=)"
        )

    # Names that cannot work as a bare string, because the connection details
    # live in the object. Catch it now rather than at first use.
    if tools_name is not None:
        try:
            from ..managed._sandbox_adapter import NEEDS_INSTANCE
        except Exception:
            NEEDS_INSTANCE = {}
        if tools_name in NEEDS_INSTANCE:
            raise TypeError(
                f"{owner}(tools_run_on={tools_run_on!r}) needs more than a name: "
                f"{NEEDS_INSTANCE[tools_name]}"
            )

    # ── two names for the same thing ─────────────────────────────────────────
    if run_on is not None and backend is not None:
        raise TypeError(
            f"{owner}(run_on=..., backend=...) sets the agent's runtime twice. "
            f"run_on={run_on!r} is the short form of backend=HostedAgent("
            f"provider={run_on!r}); pass one or the other.\n"
            f"  Use backend= only when you need to configure the runtime "
            f"(model, system prompt, tools)."
        )

    if run_on is not None and tools_run_on is not None:
        raise TypeError(
            f"{owner}(run_on={run_on!r}, tools_run_on={tools_run_on!r}) points "
            f"the tools at two machines. run_on= already runs everything -- "
            f"including tools -- on {run_on!r}.\n"
            f"  Whole agent remote:  {owner}(run_on={run_on!r})\n"
            f"  Only tools remote:   {owner}(tools_run_on={tools_run_on!r})"
        )

    if backend is not None and tools_run_on is not None:
        raise TypeError(
            f"{owner}(backend=..., tools_run_on={tools_run_on!r}) points the "
            f"tools at two machines: backend= runs the whole agent (tools "
            f"included) on its managed runtime, so the tools cannot also run on "
            f"{tools_run_on!r}.\n"
            f"  Drop tools_run_on= to keep the hosted runtime, or drop backend= "
            f"to keep local thinking with remote tools."
        )

    # ── a class that cannot answer this question ─────────────────────────────
    if run_on is not None and not supports_run_on:
        raise TypeError(
            f"{owner}(run_on={run_on!r}) is not supported: run_on= hands one "
            f"agent's whole loop to a managed runtime, and a {owner} "
            f"orchestrates several agents locally.\n"
            f"  To run every step's tools in one shared sandbox:\n"
            f"      {owner}(tools_run_on={run_on!r})\n"
            f"  To host a single agent's loop, put run_on= on that Agent."
        )

    return Placement(whole=run_on, tools=tools_run_on)

File: src/praisonai-ts/src/agent/features/placement.ts
/**
 * "Where does this run?" — `backend=`, `runOn=` and `toolsRunOn=`
 * (Python parity: `agent/placement.py`, `agent/tools_placement.py`,
 * `agent/execution_mixin.py::_delegate_to_backend`).
 *
 * Two different questions, deliberately kept apart:
 *
 * - `runOn` / `backend` place the **whole agent** — model calls, loop and
 *   tools — on a managed runtime. `runOn: "anthropic"` is shorthand for
 *   `backend: <hosted agent for anthropic>`.
 * - `toolsRunOn` places **only the tools**. Thinking stays here; the shell,
 *   the file writes and the code execution move.
 *
 * Because the two answer sets are disjoint, a wrong pairing is a typo worth
 * catching rather than a preference worth honouring: `runOn: "docker"` cannot
 * mean anything, and gets an error naming the parameter the caller wanted.
 * That validation is ported verbatim from `resolve_placement`.
 *
 * Neither a hosted runtime nor a remote sandbox ships in this package (they
 * are provider integrations), so both are resolved through registries a host
 * fills in — {@link registerManagedRuntime} and {@link registerToolPlace}.
 * An unregistered name fails loudly at construction with the same
 * "install the integration" shape Python uses, instead of running the agent
 * on the wrong machine.
 */

/** A managed runtime that can host an entire agent loop (Python `ManagedBackendProtocol`). */
export interface ManagedBackendLike {
  /** Run one prompt on managed infrastructure and return the whole response. */
  execute(prompt: unknown, options?: Record<string, unknown>): Promise<unknown> | unknown;
  /** Optional streaming form, used when the turn is streaming. */
  stream?(prompt: unknown, options?: Record<string, unknown>): AsyncIterable<string>;
  /** Optional id of the managed session, linked into the agent's session store. */
  managedSessionId?: string;
}

/**
 * A place that can run a tool call (Python's `ToolPlace` / `SharedCompute`
 * adapters). `runTool` receives the tool name, its parsed arguments and the
 * local implementation, and returns the result — a local place just calls it.
 */
export interface ToolPlaceLike {
  readonly placeName: string;
  runTool(
    toolName: string,
    args: Record<string, unknown>,
    localImplementation: () => Promise<unknown>
  ): Promise<unknown>;
}

/** The `local` place: the tool runs in this process, unchanged (Python's default). */
export const LOCAL_TOOL_PLACE: ToolPlaceLike = Object.freeze({
  placeName: 'local',
  async runTool(
    _toolName: string,
    _args: Record<string, unknown>,
    localImplementation: () => Promise<unknown>
  ): Promise<unknown> {
    return localImplementation();
  },
});

const managedRuntimes = new Map<string, () => ManagedBackendLike>();
const toolPlaces = new Map<string, () => ToolPlaceLike>([['local', () => LOCAL_TOOL_PLACE]]);

/**
 * Register a runtime that `runOn: "<name>"` may name (Python's backend
 * entry-point registry, which the wrapper package fills in).
 */
export function registerManagedRuntime(name: string, factory: () => ManagedBackendLike): void {
  managedRuntimes.set(name.toLowerCase(), factory);
}

/** Forget a managed runtime registration. Returns whether one was removed. */
export function unregisterManagedRuntime(name: string): boolean {
  return managedRuntimes.delete(name.toLowerCase());
}

/** Every runtime name `runOn=` accepts right now. */
export function managedRuntimeNames(): string[] {
  return [...managedRuntimes.keys()].sort();
}

/** Register a place that `toolsRunOn: "<name>"` may name (Python's compute bridge). */
export function registerToolPlace(name: string, factory: () => ToolPlaceLike): void {
  toolPlaces.set(name.toLowerCase(), factory);
}

/** Forget a tool place registration. `local` cannot be removed. */
export function unregisterToolPlace(name: string): boolean {
  if (name.toLowerCase() === 'local') return false;
  return toolPlaces.delete(name.toLowerCase());
}

/** Every place name `toolsRunOn=` accepts right now. */
export function toolPlaceNames(): string[] {
  return [...toolPlaces.keys()].sort();
}

/** Anything with an `execute()` is a backend (Python: `hasattr(backend, 'execute')`). */
export function isManagedBackendLike(value: unknown): value is ManagedBackendLike {
  if (value === null || (typeof value !== 'object' && typeof value !== 'function')) return false;
  return typeof (value as Record<string, unknown>).execute === 'function';
}

function isToolPlaceLike(value: unknown): value is ToolPlaceLike {
  if (value === null || typeof value !== 'object') return false;
  const v = value as Record<string, unknown>;
  return typeof v.runTool === 'function';
}

/** Python `_name_of`: only bare strings are routed by name; objects carry their own config. */
function nameOf(value: unknown): string | null {
  return typeof value === 'string' ? value.toLowerCase() : null;
}

function quoteList(names: readonly string[]): string {
  return names.length === 0 ? '(none registered)' : names.map((n) => `'${n}'`).join(', ');
}

/** The resolved answer to "where does this run?" (Python `Placement`). */
export interface Placement {
  /** The managed backend running the whole agent, or `undefined`. */
  backend?: ManagedBackendLike;
  /** The place this agent's tools run, or `undefined` for in-process. */
  toolPlace?: ToolPlaceLike;
}

export interface ResolvePlacementInput {
  /** Class name, used only so errors read naturally (Python's `owner`). */
  owner?: string;
  runOn?: unknown;
  toolsRunOn?: unknown;
  backend?: unknown;
}

/**
 * Validate the three placement options against each other and resolve them
 * (Python `resolve_placement` plus `Agent._hosted_backend_for`).
 *
 * Throws when two options name two different places for the same thing, or
 * when a place is asked to do a job it cannot do.
 */
export function resolvePlacement(input: ResolvePlacementInput): Placement {
  const owner = input.owner ?? 'Agent';
  const { runOn, toolsRunOn, backend } = input;
  const managed = managedRuntimeNames();
  const places = toolPlaceNames();

  // ── a place that cannot do the job asked of it ──────────────────────────
  const runOnName = nameOf(runOn);
  if (runOn !== undefined && runOn !== null && runOnName !== null && !managedRuntimes.has(runOnName)) {
    if (toolPlaces.has(runOnName)) {
      throw new TypeError(
        `${owner}(runOn: '${runOnName}') is not valid: runOn places the whole agent -- model calls, ` +
        `loop and tools -- on a managed runtime, and '${runOnName}' runs commands but cannot host an agent loop.\n` +
        `  To run only the tools there:  ${owner}({ toolsRunOn: '${runOnName}' })\n` +
        `  Runtimes that host a whole agent: ${quoteList(managed)}`
      );
    }
    throw new TypeError(
      `${owner}(runOn: '${runOnName}') is not a known managed runtime.\n` +
      `  Runtimes that host a whole agent: ${quoteList(managed)}\n` +
      `  Places that can run tools:        ${quoteList(places)} (pass those as toolsRunOn)\n` +
      `  A managed runtime is a provider integration: register one with registerManagedRuntime('${runOnName}', ...).`
    );
  }

  const toolsName = nameOf(toolsRunOn);
  if (toolsName !== null && managedRuntimes.has(toolsName) && !toolPlaces.has(toolsName)) {
    throw new TypeError(
      `${owner}(toolsRunOn: '${toolsName}') is not valid: '${toolsName}' hosts an entire agent, not individual tools.\n` +
      `  Did you mean:  ${owner}({ runOn: '${toolsName}' })`
    );
  }

  // A typo here used to survive construction and only surface on the first
  // model turn -- after the prompt was built and, for a real run, after the
  // API spend. Validate at the call site instead.
  if (toolsName !== null && !toolPlaces.has(toolsName)) {
    throw new TypeError(
      `${owner}(toolsRunOn: '${toolsName}') is not a known place.\n` +
      `  Places that can run tools: ${quoteList(places)}\n` +
      `  Runtimes that host a whole agent: ${quoteList(managed)} (pass those as runOn)\n` +
      `  A remote place is a provider integration: register one with registerToolPlace('${toolsName}', ...).`
    );
  }

  // ── two names for the same thing ────────────────────────────────────────
  if (runOn !== undefined && runOn !== null && backend !== undefined && backend !== null) {
    throw new TypeError(
      `${owner}(runOn, backend) sets the agent's runtime twice. runOn is the short form of a hosted ` +
      `backend for that provider; pass one or the other.\n` +
      `  Use backend only when you need to configure the runtime (model, system prompt, tools).`
    );
  }

  if (runOn !== undefined && runOn !== null && toolsRunOn !== undefined && toolsRunOn !== null) {
    throw new TypeError(
      `${owner}(runOn, toolsRunOn) points the tools at two machines. runOn already runs everything ` +
      `-- including tools -- on the managed runtime.\n` +
      `  Whole agent remote:  ${owner}({ runOn })\n` +
      `  Only tools remote:   ${owner}({ toolsRunOn })`
    );
  }

  if (backend !== undefined && backend !== null && toolsRunOn !== undefined && toolsRunOn !== null) {
    throw new TypeError(
      `${owner}(backend, toolsRunOn) points the tools at two machines: backend runs the whole agent ` +
      `(tools included) on its managed runtime, so the tools cannot also run elsewhere.\n` +
      `  Drop toolsRunOn to keep the hosted runtime, or drop backend to keep local thinking with remote tools.`
    );
  }

  // ── resolve ─────────────────────────────────────────────────────────────
  const placement: Placement = {};

  if (backend !== undefined && backend !== null) {
    if (!isManagedBackendLike(backend)) {
      throw new TypeError(
        `${owner}(backend) does not support execute(): a managed backend must expose ` +
        `execute(prompt, options) returning the response.`
      );
    }
    placement.backend = backend;
  } else if (runOn !== undefined && runOn !== null) {
    if (isManagedBackendLike(runOn)) {
      placement.backend = runOn;
    } else {
      placement.backend = managedRuntimes.get(runOnName as string)!();
    }
  }

  if (toolsRunOn !== undefined && toolsRunOn !== null) {
    if (isToolPlaceLike(toolsRunOn)) {
      placement.toolPlace = toolsRunOn;
    } else if (toolsName !== null) {
      placement.toolPlace = toolPlaces.get(toolsName)!();
    } else {
      throw new TypeError(
        `${owner}(toolsRunOn) must be a place name or an object with runTool(name, args, run).`
      );
    }
  }

  return placement;
}

/** Everything Python forwards to a managed backend's `execute()`. */
export interface BackendDelegationOptions {
  temperature?: number;
  tools?: unknown;
  outputJson?: unknown;
  outputPydantic?: unknown;
  reasoningSteps?: boolean;
  stream?: boolean;
  taskName?: string;
  taskDescription?: string;
  taskId?: string;
  /** Python's `config` keyword: forwarded verbatim, never interpreted here. */
  config?: Record<string, unknown>;
  forceRetrieval?: boolean;
  skipRetrieval?: boolean;
  attachments?: readonly string[];
  toolChoice?: string;
}

/**
 * Hand the whole turn to the managed backend (Python
 * `_delegate_to_backend`). Streams through `backend.stream()` when the turn is
 * streaming and the backend offers one, so tokens still reach `onToken`.
 */
export async function delegateToBackend(
  backend: ManagedBackendLike,
  prompt: string,
  options: BackendDelegationOptions,
  onToken?: (token: string) => void
): Promise<string> {
  if (options.stream && typeof backend.stream === 'function') {
    let accumulated = '';
    for await (const chunk of backend.stream(prompt, { ...options })) {
      const text = typeof chunk === 'string' ? chunk : String(chunk);
      if (text.length === 0) continue;
      accumulated += text;
      onToken?.(text);
    }
    return accumulated;
  }
  const result = await backend.execute(prompt, { ...options });
  return result === null || result === undefined ? '' : String(result);
}
